ODIN_II/SRC/MixingOptimization.cpp

```diff
@@ -0,0 +1,161 @@
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "MixingOptimization.hpp"
+
+#include <stdint.h> // INT_MAX
+#include <vector>
+
+#include "netlist_statistic.h" // mixing_optimization_stats
+#include "multipliers.h"       // instantiate_simple_soft_multiplier
+#include "odin_error.h"        // error_message
+
+void MixingOpt::scale_counts() {
+    if (this->_blocks_count < 0 || this->_blocks_count == INT_MAX || this->_ratio < 0.0 || this->_ratio > 1.0) {
+        error_message(NETLIST, unknown_location, "%s",
+                      "The parameters for optimization kind:%i are configured incorrectly : count %i, ratio %f\n", this->_kind, this->_blocks_count, this->_ratio);
+        exit(0);
+    }
+    this->_blocks_count = this->_blocks_count * this->_ratio;
+}
+
+void MixingOpt::assign_weights(netlist_t* /*netlist*/, std::vector<nnode_t*> /*nodes*/) {
+    // compute weights for all noted nodes
+    error_message(NETLIST, unknown_location, "%s",
+                  "Assign_weights mixing optimization was called for optimization without specification provided, for kind  %i\n", this->_kind);
+    exit(0);
+}
+
+void MixingOpt::perform(netlist_t*, std::vector<nnode_t*>&) {
+    error_message(NETLIST, unknown_location, "%s",
+                  "Performing mixing optimization was called for optimization without method provided, for kind  %i\n", this->_kind);
+    exit(0);
+}
+
+MultsOpt::MultsOpt(int _exact)
+    : MixingOpt(1.0, MULTIPLY) {
+    this->_blocks_count = _exact;
+    this->_enabled = true;
+}
+
+MultsOpt::MultsOpt(float ratio)
+    : MixingOpt(ratio, MULTIPLY) {
+    if (ratio < 0.0 || ratio > 1.0) {
+        error_message(NETLIST, unknown_location, "%s",
+                      "Miltipliers mixing optimization is started with wrong ratio %f\n", ratio);
+        exit(0);
+    }
+
+    //Explicitly set all hard block multipliers to max
+    this->_blocks_count = INT_MAX;
+    this->_enabled = true;
+}
+
+bool MultsOpt::enabled(nnode_t* /*node*/) {
+    return this->_enabled;
+}
+
+bool MultsOpt::hardenable(nnode_t* node) {
+    return (hard_multipliers && (node->input_port_sizes[0] > 1) && (node->input_port_sizes[1] > 1));
```

Comment:
can you replace "1" with min_mult from `partial_map.cpp`? its probably part of config_t if I'm not worng

Reply:
done, thank you. Will ship the changes with the next push